src/Metadata/PropertyTypeIterable.php

```diff
@@ -4,67 +4,105 @@
+ * This property type can be merged with PropertyTypeClass<T>, provided that T is, inherits from, or is a parent class of {@see PropertyTypeIterable::traversableClass}
  * This property type can be merged with PropertyTypeIterable, if :
  *  - we're not merging a plain array PropertyTypeIterable into a hashmap one,
- *  - and the collection classes of each are either not present on both sides, or are the same, or parent-child of one another
+ *  - and the traversable classes of each are either not present on either sides, or are the same, or parent-child of one another
  */
-final class PropertyTypeIterable extends PropertyTypeArray
+final class PropertyTypeIterable extends AbstractPropertyType
 {
+    /**
+     * @var PropertyType
+     */
+    private $subType;
+
+    /**
+     * @var bool
+     */
+    private $hashmap;
+
     /**
      * @var string
      */
-    private $collectionClass;
+    private $traversableClass;
 
     /**
-     * @param class-string<\Traversable>|null $collectionClass
+     * @param class-string<\Traversable>|null $traversableClass
      */
-    public function __construct(PropertyType $subType, bool $hashmap, bool $nullable, string $collectionClass = null)
+    public function __construct(PropertyType $subType, bool $hashmap, bool $nullable, string $traversableClass = null)
     {
-        parent::__construct($subType, $hashmap, $nullable, null != $collectionClass);
+        parent::__construct($nullable);
 
-        $this->collectionClass = $collectionClass;
+        $this->subType = $subType;
+        $this->hashmap = $hashmap;
+        $this->traversableClass = $traversableClass;
     }
 
     public function __toString(): string
     {
         if ($this->subType instanceof PropertyTypeUnknown) {
-            return 'array'.($this->isCollection() ? '|\\'.$this->collectionClass : '');
+            return 'array'.($this->hasTraversableClass() ? '|\\'.$this->traversableClass : '');
         }
 
         $array = $this->isHashmap() ? '[string]' : '[]';
-        if ($this->isCollection()) {
+        if ($this->hasTraversableClass()) {
             $collectionType = $this->isHashmap() ? ', string' : '';
-            $array .= sprintf('|\\%s<%s%s>', $this->collectionClass, $this->subType, $collectionType);
+            $array .= sprintf('|\\%s<%s%s>', $this->traversableClass, $this->subType, $collectionType);
         }
 
-        return ((string) $this->subType).$array.AbstractPropertyType::__toString();
+        return ((string) $this->subType).$array.parent::__toString();
+    }
+
+    public function isHashmap(): bool
+    {
+        return $this->hashmap;
+    }
+
+    /**
+     * Returns the type of the next level, which could be an array or hashmap or another type.
+     */
+    public function getSubType(): PropertyType
+    {
+        return $this->subType;
     }
 
-    public function getCollectionClass(): ?string
+    public function getTraversableClass(): ?string
```

Comment:
because we have the `hasTraversableClass` method, i think we should actually make this method only return `string` and throw an exception if it is called and traversableClass is null